Refs #35530 - Even safer save_to_file

adamruzicka opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

ianballou reviewed (dismissed):
This fixed https://github.com/theforeman/foreman/pull/9430 for me!

ianballou commented:
[test katello]

ekohl commented on app/services/foreman/renderer/scope/macros/base.rb:
```diff
@@ -112,12 +112,8 @@ def pxe_kernel_options
-              "cat << EOF | sed 's/^ //' > #{filename}\n#{content}\nEOF"
-            end
+            encoded = Base64.encode64(content || '')
+            "cat << EOF | base64 -d > #{filename}\n#{encoded}EOF"
```
This relies on `Base64.encode64` returning with a line ending, but on an empty string it doesn't:
```
[1] pry(main)> require 'base64'
=> true
[2] pry(main)> Base64.encode64('bla')
=> "Ymxh\n"
[3] pry(main)> Base64.encode64('')
=> ""
```

Perhaps you could generate `echo > #{filename}` if `encoded.empty?` to make it very obvious the file is being emptied.

ekohl commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Kickstart_default_finish.host4dhcp.snap.txt:
```diff
@@ -78,12 +78,11 @@ echo "Performing initial puppet run for --tags no_such_tag"
- /usr/bin/curl -v -k -s --data "host_config_key=" https:///api/v2/job_templates//callback/
- echo "DONE"
+cat << EOF | base64 -d > /root/ansible_provisioning_call.sh
+IyEvYmluL3NoCgplY2hvICJDYWxsaW5nIEFuc2libGUgQVdYL1Rvd2VyIHBy
```
It does make it harder to review what's really going on.

adamruzicka replied:
That's true. I wouldn't mind doing it differently. Any suggestions?

ekohl replied:
I'm looking at the original issue and wonder what's going on. One possible solution is to use a different string than `EOF` (given it's quite common). For example `FOREMAN_WRITE_CONTENT`. It reduces the risk.

Another thing to consider is that you may actually change functionality. Because it's base64 that means bash won't replace variables. It can be that something relies on that functionality.

Other solutions that come to mind are making base64 behavior opt-in. So `save_to_file(filename, content, base64: false)`.

A third option that is specific to REX setup is skip the whole write-to-file-and-then-execute-that-file approach. Just run it literally. You are already in a shell, right? Or is it intentional that you want to leave the script on disk?

adamruzicka replied:
> For example FOREMAN_WRITE_CONTENT. It reduces the risk

Reduces, but does not remove it completely. How about having it a bit dynamic and adding a, possibly random, suffix to it unless we arrive at a delimiter that is not present in the thing being saved?

> Because it's base64 that means bash won't replace variables. It can be that something relies on that functionality.

To me the variable replacing behaviour was surprising, but sadly, it's possible that something somewhere depends on it. How about a separate helper or a flag to signify if the content should be placed to the disk literally?

>  You are already in a shell, right?

Well, yes, but the same template is used in different contexts (rex job, provisioning templates), in some of them running it in a separate shell might be desirable (the rex pull setup snippet may call exit in some cases)

ekohl replied:
> How about having it a bit dynamic and adding a, possibly random, suffix to it unless we arrive at a delimiter that is not present in the thing being saved?

Random will mess with the snapshots. You could hash the filename and/or content though and add that to it.

> To me the variable replacing behaviour was surprising, but sadly, it's possible that something somewhere depends on it. How about a separate helper or a flag to signify if the content should be placed to the disk literally?

That's kind of what my second suggestion would do. Whether you name it `base64` or something else can be considered an implementation detail.

> Well, yes, but the same template is used in different contexts (rex job, provisioning templates), in some of them running it in a separate shell might be desirable (the rex pull setup snippet may call exit in some cases)

That's a fair point.

ekohl reviewed (commented):
I think it's safe to assume base64 is present:
Debian:
```console
$ dpkg -S `which base64`
coreutils: /usr/bin/base64
```

Fedora:
```
$ rpm -qf `which base64`
coreutils-9.0-8.fc36.x86_64
```

I don't know about the BSDs.

adamruzicka commented:
> I don't know about the BSDs.

{Net,Free}BSD have it, OpenBSD has `uuencode` with some flags instead. Is that a show stopper?

ekohl commented:
> {Net,Free}BSD have it, OpenBSD has `uuencode` with some flags instead. Is that a show stopper?

Not for me. It would be nice to note that limitation in the commit message so anyone looking at the history could easily figure it out using git. I'd also appreciate if you could quickly describe the issue you're solving in the commit message so we don't need to open Redmine when using `git log`.

adamruzicka committed:
Refs #35530 - Alternative approach to save_to_file

save_to_macro generates a command using cat and a heredoc string.
However, if the content to be written contains a heredoc using the same
delimiter, save_to_macro's heredoc will be terminated by it, leading to
mangled parts of the original content and save_to_macro being left
scattered around.

Another issue is, that when the content to be written contains things
like shell variables or subprocesses, those get evaluated when the
content is being written out. This may be desirable in some cases and
surprising in others.

In this commit:
- delimiter suffix is based on a hash of the file being written
- be more explicit about when the file is truncated
- Allow optionally storing the content explicitly, without letting the
  shell replace variables in it. Note, this requires base64 utility to
  be available on the target machine.
- make the macro work with files with special characters in their names

ekohl commented on app/services/foreman/renderer/scope/macros/base.rb:
```diff
@@ -108,14 +108,21 @@ def pxe_kernel_options
+              "cat << #{delimiter} | base64 -d > #{filename}\n#{content}#{delimiter}"
             else
-              "cat << EOF > #{filename}\n#{content}\nEOF"
+              content = content + "\n" unless content.end_with?("\n")
```
```suggestion
              content += "\n" unless content.end_with?("\n")
```

ekohl reviewed (commented):
I'm happy with the design, but there's a merge conflict and the test are red.

adamruzicka commented:
:green_apple:

ekohl commented on app/services/foreman/renderer/scope/macros/base.rb:
```diff
@@ -108,15 +108,21 @@ def pxe_kernel_options
               a line end, if not an extra trailing line end is appended automatically."
             required :filename, String, desc: 'the file path to store the content to'
             required :content, String, desc: 'content to be stored'
+            keyword :explicit, [true, false], desc: 'Controls whether the file should be put on disk as-is or if variables should be replaced by shell before the file is written out', default: false
```
Thoughts on `verbatim` instead of `explicit`?

ekohl reviewed (approved):
CI is still running.

adamruzicka commented:
Not anymore :green_apple:
